src/NuGet.Core/NuGet.DependencyResolver/SourceRepositoryDependencyProvider.cs

```diff
@@ -139,9 +140,21 @@ private IEnumerable<LibraryDependency> GetDependencies(FindPackageByIdDependency
         {
             if (_findPackagesByIdResource == null)
             {
-                _findPackagesByIdResource = await _sourceRepository.GetResourceAsync<FindPackageByIdResource>();
-                _findPackagesByIdResource.Logger = _logger;
-                _findPackagesByIdResource.CacheContext = _cacheContext;
+                // Avoid multiple instances of the same resource
+                await _resourceLock.WaitAsync();
```

Comment:
why not have a lazy? This seems a lot more complex, is it because lazy doesn't support async?